lib/wit/package.py

```diff
@@ -1,73 +1,149 @@
 #!/usr/bin/env python3
 
-from . import gitrepo
-import os
 from pathlib import Path
-from .common import WitUserError
+import re
+import os
+import shutil
+from .gitrepo import GitRepo
+from .witlogger import getLogger
 
+log = getLogger()
 
-class NotAPackageError(WitUserError):
+
+class WitBug(Exception):
     pass
 
 
-# Make this a factory for different VCS types
 class Package:
-    """
-    Generic package type
-    """
 
-    @staticmethod
-    def from_arg(s) -> 'gitrepo.GitRepo':  # String fixes circular dependency
-        """
-        >>> Package.from_arg(".::HEAD")
-        gitrepo.GitRepo(source='.', revision='HEAD')
-        >>> Package.from_arg("not-a-repo")
-        Traceback (most recent call last):
-            ...
-        argparse.ArgumentTypeError: Remote git repo 'not-a-repo' does not exist!
+    def __init__(self, name, source, unresolved_revision, repo_paths):
+        """Create a package, cloning it to the .wit folder"""
+        self.name = name
+        self.source = source
+        self.unresolved_revision = unresolved_revision or "HEAD"
+        self.revision = None
+
+        self.repo = None
+
+        self.find_source(repo_paths)
+
+        self.dependents = []
+
+    def short_revision(self):
+        rev = self.revision or self.unresolved_revision
+        if len(rev) >= 40:
+            rev = rev[:8]
```

Comment:
8 is too short, it looks like 12-characters is a better number for short revisions https://blog.cuviper.com/2013/11/10/how-short-can-git-abbreviate/

Reply:
We can also just let git tell us because it guarantees uniqueness (although that doesn't guarantee anything about future collisions)

Reply:
Good catch. Fixed by https://github.com/sifive/wit/pull/96/commits/25c8bc59995342748ddce096d12abd6ce7925352